### PR TITLE
feat: Upstream Spice AI changes to `datafusion-federation`

### PR DESCRIPTION
**Pull request by phillipleblanc**
This PR contains all of the changes that the spiceai project made to `datafusion-federation` to fix bugs and add features to make federation work better for our use-cases.

The main two features that were added:

1. The name of the table registered in DataFusion doesn't need to match the name of the table in the remote database. This was an important requirement for our product, and we did a lot of work to get this working correctly.
2. The Arrow schema returned by the provider may not exactly match what DataFusion expects it to be, so we add a schema casting layer to make sure. Where possible this leverages the native Arrow cast kernels, but there are a few special cases that are handled.

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -8,12 +8,15 @@ members = [
 ]
 
 [workspace.package]
-version = "0.1.3"
+version = "0.1.6"
 edition = "2021"
-license = "MIT"
+license = "Apache-2.0"
```
Comment (phillipleblanc) [inline Cargo.toml:13]: I felt this would be better since DataFusion is an Apache project?

File: datafusion-federation/src/lib.rs [hunk 2]
```diff
@@ -16,9 +16,15 @@ use datafusion::{
...
 pub use optimizer::{get_table_source, FederationOptimizerRule};
-pub use plan_node::{FederatedPlanNode, FederatedQueryPlanner, FederationPlanner};
+pub use plan_node::{
+    FederatedPlanNode, FederatedPlanner, FederatedQueryPlanner, FederationPlanner,
+};
 pub use table_provider::{FederatedTableProviderAdaptor, FederatedTableSource};
 
+// TODO clean up this
+// TODO move schema_cast.rs to schema_cast directory
+pub mod schema_cast;
```
Comment (phillipleblanc) [inline datafusion-federation/src/lib.rs:26]: This schema_cast module is useful when the schema registered in DataFusion and the schema returned by the underlying provider don't match exactly. The schema_cast provides a mechanism to keep the schema aligned with what DataFusion expects.

File: datafusion-federation/src/sql/mod.rs [hunk 3]
```diff
@@ -100,6 +113,481 @@ impl OptimizerRule for SQLFederationOptimizerRule {
         false
     }
 }
+
+/// Rewrite table scans to use the original federated table name.
+fn rewrite_table_scans(
```
Comment (phillipleblanc) [inline datafusion-federation/src/sql/mod.rs:118]: This is needed to support the table name in the remote source being different from the table name registered in DataFusion. This required a lot of trial and error to get right, and we have a good integration test coverage of this in spiceai.

File: datafusion-federation/src/sql/mod.rs [hunk 4]
```diff
@@ -162,13 +661,18 @@ impl DisplayAs for VirtualExecutionPlan {
...
+        };
+
         write!(f, " sql={ast}")
     }
 }
 
 impl ExecutionPlan for VirtualExecutionPlan {
     fn name(&self) -> &str {
-        "sql_federation_exec"
+        "VirtualExecutionPlan"
```
Comment (phillipleblanc) [inline datafusion-federation/src/sql/mod.rs:675]:
```suggestion
        "sql_federation_exec"
```

File: datafusion-federation/src/lib.rs (hunk 2 above)
Comment (hozan23) [reply datafusion-federation/src/lib.rs:26]: I put `TODO` as reminder there to clean up the module by moving schema_cast.rs to schema_cast/mod.rs

**Review by backkem — APPROVED** [review]

**Comment by backkem** [comment]
The rewriting does seem like good generic functionality so I'm fine to accept it in.

**Pull request merged**